File: output/python/tests/06_domain_vars.py
from typing import Any, Optional, List, Dict, Callable

class DomainVarsFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters


class DomainVarsFrameContext:
    def __init__(self, event: DomainVarsFrameEvent, default_return):
        self.event = event
        self._return = default_return
        self._data = {}


class DomainVarsCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'DomainVarsCompartment':
        c = DomainVarsCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class DomainVars:
    def __init__(self):
        self._state_stack = []
        self._context_stack = []
        self.count = 0
        self.name = "counter"
        self.__compartment = DomainVarsCompartment("Counting")
        self.__next_compartment = None
        __frame_event = DomainVarsFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = DomainVarsFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = DomainVarsFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = DomainVarsFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def increment(self):
        __e = DomainVarsFrameEvent("increment", None)
        __ctx = DomainVarsFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def decrement(self):
        __e = DomainVarsFrameEvent("decrement", None)
        __ctx = DomainVarsFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def get_count(self) -> int:
        __e = DomainVarsFrameEvent("get_count", None)
        __ctx = DomainVarsFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        return self._context_stack.pop()._return

    def set_count(self, value: int):
        __e = DomainVarsFrameEvent("set_count", {"value": value})
        __ctx = DomainVarsFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def _state_Counting(self, __e):
        if __e._message == "decrement":
            self.count -= 1
            print(f"{self.name}: decremented to {self.count}")
        elif __e._message == "get_count":
            self._context_stack[-1]._return = self.count
            return
        elif __e._message == "increment":
            self.count += 1
            print(f"{self.name}: incremented to {self.count}")
        elif __e._message == "set_count":
            value = __e._parameters["value"]
            self.count = value
            print(f"{self.name}: set to {self.count}")


def main():
    print("=== Test 06: Domain Variables ===")
    s = DomainVars()

    # Initial value should be 0
    count = s.get_count()
    assert count == 0, f"Expected initial count=0, got {count}"
    print(f"Initial count: {count}")

    # Increment
    s.increment()
    count = s.get_count()
    assert count == 1, f"Expected count=1, got {count}"

    s.increment()
    count = s.get_count()
    assert count == 2, f"Expected count=2, got {count}"

    # Decrement
    s.decrement()
    count = s.get_count()
    assert count == 1, f"Expected count=1, got {count}"

    # Set directly
    s.set_count(100)
    count = s.get_count()
    assert count == 100, f"Expected count=100, got {count}"

    print(f"Final count: {count}")
    print("PASS: Domain variables work correctly")

if __name__ == '__main__':
    main()

File: output/typescript/tests/06_domain_vars.ts
class DomainVarsFrameEvent {
    public _message: string;
    public _parameters: Record<string, any> | null;

    constructor(message: string, parameters: Record<string, any> | null) {
        this._message = message;
        this._parameters = parameters;
    }
}


class DomainVarsFrameContext {
    public event: DomainVarsFrameEvent;
    public _return: any;
    public _data: Record<string, any>;

    constructor(event: DomainVarsFrameEvent, default_return: any) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


class DomainVarsCompartment {
    public state: string;
    public state_args: Record<string, any>;
    public state_vars: Record<string, any>;
    public enter_args: Record<string, any>;
    public exit_args: Record<string, any>;
    public forward_event: any;
    public parent_compartment: DomainVarsCompartment | null;

    constructor(state: string, parent_compartment: DomainVarsCompartment | null = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    public copy(): DomainVarsCompartment {
        const c = new DomainVarsCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


class DomainVars {
    private _state_stack: Array<any>;
    private __compartment: DomainVarsCompartment;
    private __next_compartment: DomainVarsCompartment | null;
    private _context_stack: Array<any>;
    private count: number = 0;
    private name: string = "counter";

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.count = 0;
        this.name = "counter";
        this.__compartment = new DomainVarsCompartment("Counting");
        this.__next_compartment = null;
        const __frame_event = new DomainVarsFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    private __kernel(__e: DomainVarsFrameEvent) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new DomainVarsFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new DomainVarsFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new DomainVarsFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    private __router(__e: DomainVarsFrameEvent) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    private __transition(next_compartment: DomainVarsCompartment) {
        this.__next_compartment = next_compartment;
    }

    public increment() {
        const __e = new DomainVarsFrameEvent("increment", null);
        const __ctx = new DomainVarsFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    public decrement() {
        const __e = new DomainVarsFrameEvent("decrement", null);
        const __ctx = new DomainVarsFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    public get_count(): number {
        const __e = new DomainVarsFrameEvent("get_count", null);
        const __ctx = new DomainVarsFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()!._return;
    }

    public set_count(value: number) {
        const __e = new DomainVarsFrameEvent("set_count", {"value": value});
        const __ctx = new DomainVarsFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    private _state_Counting(__e: DomainVarsFrameEvent) {
        if (__e._message === "decrement") {
            this.count -= 1;
            console.log(`${this.name}: decremented to ${this.count}`);
        } else if (__e._message === "get_count") {
            this._context_stack[this._context_stack.length - 1]._return = this.count;
            return;;
        } else if (__e._message === "increment") {
            this.count += 1;
            console.log(`${this.name}: incremented to ${this.count}`);
        } else if (__e._message === "set_count") {
            const value = __e._parameters?.["value"];
            this.count = value;
            console.log(`${this.name}: set to ${this.count}`);
        }
    }
}


function main() {
    console.log("=== Test 06: Domain Variables ===");
    const s = new DomainVars();

    // Initial value should be 0
    let count = s.get_count();
    if (count !== 0) {
        throw new Error(`Expected initial count=0, got ${count}`);
    }
    console.log(`Initial count: ${count}`);

    // Increment
    s.increment();
    count = s.get_count();
    if (count !== 1) {
        throw new Error(`Expected count=1, got ${count}`);
    }

    s.increment();
    count = s.get_count();
    if (count !== 2) {
        throw new Error(`Expected count=2, got ${count}`);
    }

    // Decrement
    s.decrement();
    count = s.get_count();
    if (count !== 1) {
        throw new Error(`Expected count=1, got ${count}`);
    }

    // Set directly
    s.set_count(100);
    count = s.get_count();
    if (count !== 100) {
        throw new Error(`Expected count=100, got ${count}`);
    }

    console.log(`Final count: ${count}`);
    console.log("PASS: Domain variables work correctly");
}

main();
